subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CacheStep.java

```diff
@@ -17,121 +17,148 @@
+                }))
+            )
+            .orElseGet(() -> executeWithCache(cacheHandler, context));
     }
 
-    @Nullable
-    private BuildCacheCommandFactory.LoadMetadata load(UnitOfWork work, BuildCacheKey cacheKey) {
+    private Optional<Try<BuildCacheCommandFactory.LoadMetadata>> load(UnitOfWork work, BuildCacheKey cacheKey) {
         try {
-            return buildCache.load(
-                    commandFactory.createLoad(cacheKey, work, new BuildCacheLoadListener() {
-                        @Override
-                        public void beforeLoad() {
-                            Optional<? extends Iterable<String>> changingOutputs = work.getChangingOutputs();
-                            changingOutputs.ifPresent(affectedPaths -> outputChangeListener.beforeOutputChange(affectedPaths));
-                            if (!changingOutputs.isPresent()) {
-                                outputChangeListener.beforeOutputChange();
-                            }
-                        }
-
-                        @Override
-                        public void afterLoadFailedAndWasCleanedUp(Throwable error) {
-                            work.outputsRemovedAfterFailureToLoadFromCache();
-                        }
-                    })
-            );
-        } catch (UnrecoverableUnpackingException e) {
-            // We didn't manage to recover from the unpacking error, there might be leftover
-            // garbage among the task's outputs, thus we must fail the build
-            throw e;
+            return buildCache.load(commandFactory.createLoad(cacheKey, work))
+                .map(loadResult -> Try.successful(loadResult));
         } catch (Exception e) {
             // There was a failure during downloading, previous task outputs should be unaffected
-            LOGGER.warn("Failed to load cache entry for {}, falling back to executing task", work.getDisplayName(), e);
-            return null;
+            LOGGER.warn("Failed to load cache entry for {}, cleaning outputs and falling back to (non-incremental) execution", work.getDisplayName(), e);
+            cleanupTreesAfterUnpackFailure(work);
+            return Optional.of(Try.failure(e));
         }
     }
 
+    private static void cleanupTreesAfterUnpackFailure(UnitOfWork work) {
+        work.visitOutputTrees((name, type, root) -> {
+            try {
+                if (root.exists()) {
+                    if (root.isDirectory()) {
+                        FileUtils.cleanDirectory(root);
+                    } else {
+                        FileUtils.forceDelete(root);
+                    }
+                }
+            } catch (IOException ex) {
+                throw new UncheckedIOException(String.format("Failed to clean up files for tree '%s' of %s: %s", name, work.getDisplayName(), root), ex);
+            }
+        });
+    }
+
+    private CurrentSnapshotResult executeWithCache(CacheHandler cacheHandler, IncrementalChangesContext context) {
```

Comment:
The method name `executeWithCache` confused me a little bit. How about `executeAndStoreInCache`?

Reply:
Alright.